### ISSUE TITLE
gorm race mutex unlock bug

### ISSUE THREAD

**Issue by chnzrb** [jinzhu/gorm#2366]
// SingularTable use singular table by default
func (s *DB) SingularTable(enable bool) {
modelStructsMap = sync.Map{}
s.parent.singularTable = enable
}
Concurrency panic sync: unlock of unlocked mutex。

fatal error: sync: unlock of unlocked mutex

goroutine 8817 [running]:
runtime.throw(0xbbdcf0, 0x1e)
C:/Go/src/runtime/panic.go:617 +0x72 fp=0xc0002109e0 sp=0xc0002109b0 pc=0x42d342
sync.throw(0xbbdcf0, 0x1e)
C:/Go/src/runtime/panic.go:603 +0x35 fp=0xc000210a00 sp=0xc0002109e0 pc=0x42d2c5
sync.(*Mutex).Unlock(0x119f880)
C:/Go/src/sync/mutex.go:184 +0xc1 fp=0xc000210a28 sp=0xc000210a00 pc=0x463661
sync.(*Map).Store(0x119f880, 0xb86860, 0xb7f2e0, 0xad97a0, 0xc000216780)
C:/Go/src/sync/map.go:162 +0x269 fp=0xc000210ad0 sp=0xc000210a28 pc=0x461de9
github.com/jinzhu/gorm.(*Scope).GetModelStruct(0xc000772b80, 0x0)
F:/GOHOME/src/github.com/jinzhu/gorm/model_struct.go:615 +0x1390 fp=0xc000210d80 sp=0xc000210ad0 pc=0x8dc6f0
github.com/jinzhu/gorm.(*Scope).PrimaryField(0xc000772b80, 0xaacdc0)
F:/GOHOME/src/github.com/jinzhu/gorm/scope.go:164 +0x2f fp=0xc000210db8 sp=0xc000210d80 pc=0x8de58f
github.com/jinzhu/gorm.queryCallback(0xc000772b80)
F:/GOHOME/src/github.com/jinzhu/gorm/callback_query.go:36 +0xc6d fp=0xc000210f88 sp=0xc000210db8 pc=0x8bf32d
github.com/jinzhu/gorm.(*Scope).callCallbacks(0xc000772b80, 0xc0001d1a40, 0x3, 0x4, 0x0)
F:/GOHOME/src/github.com/jinzhu/gorm/scope.go:867 +0x88 fp=0xc000210fb8 sp=0xc000210f88 pc=0x8e7358
github.com/jinzhu/gorm.(*DB).First(0xc000236000, 0xa83980, 0xc00000ea48, 0x0, 0x0, 0x0, 0xc00000ea38)
F:/GOHOME/src/github.com/jinzhu/gorm/main.go:309 +0x128 fp=0xc000211040 sp=0xc000210fb8 pc=0x8d51f8
myadmin_h5/models.GetServerNode(...)

link：golang/go#26480

**Comment by edocevol** [comment]
The same problem .
````
goroutine 81 [running]:
runtime.throw(0xc4422c, 0x1e)
        /usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc0004f4d50 sp=0xc0004f4d20 pc=0x42eea2
sync.throw(0xc4422c, 0x1e)
        /usr/local/go/src/runtime/panic.go:603 +0x35 fp=0xc0004f4d70 sp=0xc0004f4d50 pc=0x42ee25
sync.(*Mutex).Unlock(0x1253300)
        /usr/local/go/src/sync/mutex.go:184 +0xc1 fp=0xc0004f4d98 sp=0xc0004f4d70 pc=0x473cb1
sync.(*Map).Load(0x1253300, 0xc1dd60, 0xbb8f60, 0xd777a0, 0xb72da0, 0x4044c4)
        /usr/local/go/src/sync/map.go:119 +0x1e6 fp=0xc0004f4e10 sp=0xc0004f4d98 pc=0x4720d6
github.com/jinzhu/gorm.(*Scope).GetModelStruct(0xc0009fe480, 0x0)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/model_struct.go:166 +0x21e fp=0xc0004f50c0 sp=0xc0004f4e10 pc=0xa18cbe
github.com/jinzhu/gorm.(*Scope).GetStructFields(0xc0009fe480, 0xc000a2b080, 0x16, 0xc0007c3901)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/model_struct.go:622 +0x2b fp=0xc0004f50e0 sp=0xc0004f50c0 pc=0xa1a08b
github.com/jinzhu/gorm.(*Scope).GetModelStruct.func2(0xc00099d500, 0xc0009fe000, 0xc0004bff80, 0xc0009c1590)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/model_struct.go:437 +0x192 fp=0xc0004f5368 sp=0xc0004f50e0 pc=0xa37032
github.com/jinzhu/gorm.(*Scope).GetModelStruct(0xc0009fe000, 0xc0004bff80)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/model_struct.go:617 +0x1387 fp=0xc0004f5618 sp=0xc0004f5368 pc=0xa19e27
github.com/jinzhu/gorm.(*Scope).PrimaryField(0xc0009fe000, 0xb17740)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/scope.go:164 +0x2f fp=0xc0004f5650 sp=0xc0004f5618 pc=0xa1bd1f
github.com/jinzhu/gorm.queryCallback(0xc0009fe000)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/callback_query.go:36 +0xc76 fp=0xc0004f5820 sp=0xc0004f5650 pc=0x9fcaa6
github.com/jinzhu/gorm.(*Scope).callCallbacks(0xc0009fe000, 0xc000270020, 0x3, 0x4, 0x0)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/scope.go:867 +0x88 fp=0xc0004f5850 sp=0xc0004f5820 pc=0xa24ac8
github.com/jinzhu/gorm.(*DB).First(0xc0008dfce0, 0xae7740, 0xc000714830, 0x0, 0x0, 0x0, 0xc0008dfce0)
        /mnt/e/GOOOOO/pkg/mod/github.com/jinzhu/gorm@v1.9.2/main.go:294 +0x128 fp=0xc0004f58d8 sp=0xc0004f5850 pc=0xa128e8
fast-api/app/controller.(*CompanyController).Staff(0x1273248, 0xc0003cad10)
````

**Comment by emirb** [comment]
Fixed in #2396 and released in [v1.9.4](https://github.com/jinzhu/gorm/releases/tag/v1.9.4).
